test/unit/plugins/divclass.test.js

```diff
@@ -0,0 +1,121 @@
+      expect(output).toEqual("[][0]\n\nsimple content\n\n[/][0]\n");
+    });
+
+    it('works without content - but only if separated by FOUR newlines', () => {
+      const input = "[empty]\n\n\n\n[/empty]";
+      const output = parser.sourceToRedacted(input);
+      expect(output).toEqual("[][0]\n\n[/][0]\n");
+    });
+
+    it('renders a divclass within other content', () => {
+      const input = "outside of div\n\n[divname]\n\ninside div\n\n[/divname]\n\nmore outside";
+      const output = parser.sourceToRedacted(input);
+      expect(output).toEqual("outside of div\n\n[][0]\n\ninside div\n\n[/][0]\n\nmore outside\n");
+    });
+
+    it("doesn't care about duplicate classes", () => {
+      const input = "[classname]\n\nfirst\n\n[/classname]\n\n[classname]\n\nsecond\n\n[/classname]";
+      const output = parser.sourceToRedacted(input);
+      expect(output).toEqual("[][0]\n\nfirst\n\n[/][0]\n\n[][1]\n\nsecond\n\n[/][1]\n");
+    });
+
+    it('can redact nested divclasses', () => {
+      const input = "[outer]\n\n[inner]\n\nnested\n\n[/inner]\n\n[/outer]";
+      const output = parser.sourceToRedacted(input);
+      expect(output).toEqual("[][0]\n\n[][1]\n\nnested\n\n[/][1]\n\n[/][0]\n");
+    });
+
+    it('can redact inline content inside a divclass', () => {
+      const input = "[complex-example]\n\n-   an ordered list\n-   with [other redacted content](http://example.com)\n\n[/complex-example]";
+      const output = parser.sourceToRedacted(input);
+      expect(output).toEqual("[][0]\n\n-   an ordered list\n-   with [other redacted content][1]\n\n[/][0]\n");
+    });
+  });
+
+  describe('restore', () => {
+    it('can restore basic divclasses', () => {
+      const source = "[col-33]\n\nsimple content\n\n[/col-33]";
+      const redacted = "[][0]\n\ncontenu simple\n\n[/][0]\n";
+      const output = parser.sourceAndRedactedToHtml(source, redacted);
+      expect(output).toEqual("<div class=\"col-33\"><p>contenu simple</p></div>\n");
+    });
+
+    it('can restore nested divclasses', () => {
+      const source = "[outer]\n\n[inner]\n\nnested\n\n[/inner]\n\n[/outer]";
+      const redacted = "[][0]\n\n[][1]\n\nimbriqué\n\n[/][1]\n\n[/][0]\n";
+      const output = parser.sourceAndRedactedToHtml(source, redacted);
+      expect(output).toEqual("<div class=\"outer\"><div class=\"inner\"><p>imbriqué</p></div></div>\n");
+    });
+
+    it('can restore inline content inside a divclass', () => {
+      const source = "[complex-example]\n\n-   an ordered list\n-   with [other redacted content](http://example.com)\n\n[/complex-example]";
+      const redacted = "[][0]\n\n-   une liste ordonnée\n-   avec [d'autres contenus rédigés][1]\n\n[/][0]\n";
+      const output = parser.sourceAndRedactedToHtml(source, redacted);
+      expect(output).toEqual("<div class=\"complex-example\"><ul>\n<li>une liste ordonnée</li>\n<li>avec <a href=\"http://example.com\">d'autres contenus rédigés</a></li>\n</ul></div>\n");
+    });
```

Comment:
I'd like to see more test cases around restoration when translators make surprising changes or misbehave.  Some ideas:

* Ordering changes, as we've previously discussed:
  ```
  [zero]\n\n\n\n[/zero]\n\n[one]\n\n\n\n[/one]
  [][1]\n\n\n\n[/][1]\n\n[][0]\n\n\n\n[\][0]
  ```
* Nesting changes
  ```
  [zero]\n\n[one]\n\n\n\n[/one]\n\n[/zero]
  # Un-nesting?
  [][0]\n\n\n\n[/][0]\n\n[][1]\n\n\n\n[/][1]
  # Inverted nesting?
  [][1]\n\n[][0]\n\n\n\n[/][0]\n\n[/][1]
  # Broken nesting?
  [][0]\n\n[][[1]\n\n\n\n[/][0]\n\n[/][1]
  ```
* Additional content
  ```
  [item]\n\nFirst\n\n[/item]\n\n[item]\n\nSecond\n\n[/item]
  # Reusing an index
  [][0]\n\nPremier\n\n[/][0]\n\n[][1]\n\nDeuxième\n\n[/][1]\n\n[][1]\n\nTroisième\n\n[/][1]
  # Extra indices
  [][0]\n\nPremier\n\n[/][0]\n\n[][1]\n\nDeuxième\n\n[/][1]\n\n[][2]\n\nTroisième\n\n[/][2]
  ```
* Translator removes required whitespace
  ```
  [clazz]\n\nCat\n\n[/clazz]
  [][0]\nChat\n[/][0]
  ```

Not that we have to handle all these bad cases wonderfully, but just so we know we're consistent.